### ISSUE TITLE
False positive in `map_identity` with lifetimes

### ISSUE THREAD

**Issue by human-0** [rust-lang/rust-clippy#9280]
### Summary

Mapping with an identity closure (`|x| x`) is sometimes required to fix lifetimes.

### Lint Name

map_identity

### Reproducer

I tried this code:

```rust
pub fn test<'a, I>(iter: I)
where
    I: IntoIterator<Item = &'a i32>
{
    let local = 0;
    let _ = std::iter::once(&local).chain(iter.into_iter().map(|x| x));
}
```
Removing the `map` results in a compile error.
I saw this happen:

```
warning: unnecessary map of the identity function
 --> src/lib.rs:6:59
  |
6 |     let _ = std::iter::once(&local).chain(iter.into_iter().map(|x| x));
  |                                                           ^^^^^^^^^^^ help: remove the call to `map`
  |
  = note: `#[warn(clippy::map_identity)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_identity
```

I expected to see this happen:
No warning, as the `map` cannot be removed without causing a compile error. This happens on both stable and nightly.

### Version

```text
rustc 1.64.0-nightly (0f4bcadb4 2022-07-30)
binary: rustc
commit-hash: 0f4bcadb46006bc484dad85616b484f93879ca4e
commit-date: 2022-07-30
host: x86_64-pc-windows-msvc
release: 1.64.0-nightly
LLVM version: 14.0.6
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by squ1dd13** [comment]
Does clippy have enough information to be able to ignore identity `map` calls where the type input into the closure is different from the type that it outputs? Would that be a viable solution?

**Comment by xFrednet** [comment]
> Does clippy have enough information to be able to ignore identity `map` calls where the type input into the closure is different from the type that it outputs?

Yes, Clippy has all the information that rustc has.

> Would that be a viable solution?

Also yes, the current implementation is a bit naive IIRC. That causes a false positive. Even though I'm a bit surprised that we need the `|x| x` here. :thinking:

**Comment by cyqsimon** [comment]
Just ran into this in sharkdp/bat#2755.

Asked on community Discord and got a proper explanation:

> It's not that Iterator::map is "downcasting lifetimes", it's just that impl Iterator<Item = T> is invariant in T, so you need a closure to "do" the variance

See: https://discord.com/channels/273534239310479360/1120124565591425034/1170543402870382653

**Comment by cyqsimon** [comment]
I'll give this one a shot. Never messed with the compiler/linter before so wish me luck.

@rustbot claim

**Comment by xFrednet** [comment]
Hmm, I haven't looked much into this issue, but I would suggest you to look at the following two things:
1. You can look at the HIR structure of the expression with the `#[clippy::dump]` attribute: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=00f52daacb276412d25625dd4d6fa2d8
2. For the actual solution, you'll probably have a look at the expression type of the parameter and return type. This might give you some more hints how to get the semantic type: https://doc.rust-lang.org/clippy/development/common_tools_writing_lints.html Note that lifetimes are turned into regions on the semantic level.

**Comment by Alexendoo** [comment]
Some lifetimes are erased, it may not be possible to detect

**Comment by waynexia** [comment]
Run into this with another example:
```rust
pub fn foo(
    input: impl IntoIterator<Item = &'a (A, B)>,
) {
    let _ : HashMap<&'a A, &'a B> = input.into_iter().map(|(k, v)| (k, v)).collect();
}
```

Where the input type is `&(A, B)` and output type is `(&A, &B)`

**Comment by y21** [comment]
> Where the input type is `&(A, B)` and output type is `(&A, &B)`

That bug was fixed in #11792 and I think match ergonomics is unrelated to the lifetime problem here?

If I run your code with a newer clippy build (tested nightly 2023-12-03), this indeed no longer emits a warning.

**Comment by stephaneyfx** [comment]
The bug regarding `&(A, B)` turned into `(&A, &B)` seems to be [back](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=448c6d409b88fdda5fd8f7c0ca8bf9d8).
```rust
pub struct Foo;

pub fn foo(x: Option<&(Foo, Foo)>) -> Option<(&Foo, &Foo)> {
    x.map(|(a, b)| (a, b))
}

pub fn bar(xs: Vec<&(Foo, Foo)>) -> Vec<(&Foo, &Foo)> {
    xs.into_iter().map(|(a, b)| (a, b)).collect()
}
```

Clippy output (clippy 0.1.75 (82e1608 2023-12-21)):
```
warning: unnecessary map of the identity function
 --> src/lib.rs:4:6
  |
4 |     x.map(|(a, b)| (a, b))
  |      ^^^^^^^^^^^^^^^^^^^^^ help: remove the call to `map`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_identity
  = note: `#[warn(clippy::map_identity)]` on by default

warning: unnecessary map of the identity function
 --> src/lib.rs:8:19
  |
8 |     xs.into_iter().map(|(a, b)| (a, b)).collect()
  |                   ^^^^^^^^^^^^^^^^^^^^^ help: remove the call to `map`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_identity
```

**Comment by y21** [comment]
@stephaneyfx Your Rust playground link uses the stable channel and stable does not have that fix yet (it takes quite a while for a change made today to actually land it on stable, and the date that the playground is a bit misleading). Changing the channel to nightly makes the warnings go away, so I presume that it actually is fixed?
Then again, this is different from the original issue here, so please open a new issue if you think this case isn't properly fixed

**Comment by stephaneyfx** [comment]
My bad for incorrectly assuming the two-month-old fix was available on stable. I cannot reproduce using nightly with either the playground or my actual code, as you mentioned. Thank you and sorry for the noise.
